L1Trigger/Phase2L1ParticleFlow/plugins/L1HPSPFTauProducer.cxx

```diff
@@ -0,0 +1,206 @@
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+  
+
+private:
+  
+  
+  void produce(edm::StreamID, edm::Event& iEvent, const edm::EventSetup& iSetup) const override;
+  
+
+  //various needed vars
+  int _nTaus;
+  bool _HW;
+  bool fUseJets_;
+  bool _debug;
+  
+  edm::InputTag srcL1PFCands_;
+  edm::EDGetTokenT<l1t::PFCandidateCollection> tokenL1PFCands_;    
+  //jets
+  edm::InputTag srcL1PFJets_;
+  edm::EDGetTokenT<std::vector<reco::CaloJet>> tokenL1PFJets_;
+  //functions
+  std::vector<l1t::PFTau> processEvent_HW(std::vector<edm::Ptr<l1t::PFCandidate>>& parts, std::vector<edm::Ptr<reco::CaloJet>>&  jets) const;
+
+  static std::vector<L1HPSPFTauEmu::Particle> convertJetsToHW(std::vector<edm::Ptr<reco::CaloJet>>& edmJets); 
+  static std::vector<L1HPSPFTauEmu::Particle> convertEDMToHW(std::vector<edm::Ptr<l1t::PFCandidate>>& edmParticles);
+  static std::vector<l1t::PFTau> convertHWToEDM(std::vector<L1HPSPFTauEmu::Tau> hwTaus);
+  
+};
+
+L1HPSPFTauProducer::L1HPSPFTauProducer(const edm::ParameterSet& cfg)
+   : _nTaus(cfg.getParameter<int>("nTaus")),
+     _HW(cfg.getParameter<bool>("HW")),
+     fUseJets_(cfg.getParameter<bool>("useJets")),
+     _debug(cfg.getParameter<bool>("debug")){ //,
+      srcL1PFCands_ = cfg.getParameter<edm::InputTag>("srcL1PFCands");
+      tokenL1PFCands_ = consumes<l1t::PFCandidateCollection>(srcL1PFCands_);   
+      
+      srcL1PFJets_ = cfg.getParameter<edm::InputTag>("srcL1PFJets");
+      tokenL1PFJets_ = consumes<std::vector<reco::CaloJet>>(srcL1PFJets_); 
+      produces<l1t::PFTauCollection>("HPSTaus");
+      
+      produces<l1t::PFCandidateCollection>("SelPFCands");
+
+     }
+
+void L1HPSPFTauProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions){
+  edm::ParameterSetDescription desc;    
+  
+  desc.add<edm::InputTag>("srcL1PFCands", edm::InputTag("l1tLayer1","Puppi"));    
+  desc.add<int>("nTaus", 16);
+  desc.add<bool>("HW", true);
+  desc.add<bool>("useJets",false);
+  desc.add<bool>("debug", false);   
+  desc.add<edm::InputTag>("srcL1PFJets", edm::InputTag("l1tPhase1JetProducer","UncalibratedPhase1L1TJetFromPfCandidates"));
+  descriptions.add("L1HPSPFTauProducer", desc);
+}
+
+
+void L1HPSPFTauProducer::produce(edm::StreamID,
+                                 edm::Event& iEvent,
+                                 const edm::EventSetup& iSetup) const {
+
+  std::unique_ptr<l1t::PFTauCollection> newPFTauCollection(new l1t::PFTauCollection);
+  std::unique_ptr<l1t::PFCandidateCollection> selParticles(new l1t::PFCandidateCollection);
+  edm::Handle<l1t::PFCandidateCollection> l1PFCandidates;
+  iEvent.getByToken(tokenL1PFCands_, l1PFCandidates);
+  //add jets even if not used, for simplicity
+  edm::Handle<std::vector<reco::CaloJet>> l1PFJets;
+  iEvent.getByToken(tokenL1PFJets_, l1PFJets);
+  //
+
+  //adding collection
+  std::vector<edm::Ptr<l1t::PFCandidate>> particles;
+  for(unsigned i = 0; i < (*l1PFCandidates).size(); i++) {
```

Comment:
Mayeb this is a stylistic thing on my part, but does `edm::Ptr` not allow element access with `->`?

Reply:
Using `->` here gives an error during compilation, so probably best left as is.